Add Colima installation over Docker Desktop

michaelsaki opened the pull request:
## 🗣 Description ##

This PR will remove the current Docker documentation from the [mac-env-setup.md](https://github.com/cisagov/development-guide/blob/develop/dev_envs/mac-env-setup.md) and replace it with steps to install and configure [Colima](https://github.com/abiosoft/colima).

## 💭 Motivation and context ##

Docker now requires certain users and businesses to purchase a license for use of [Docker Desktop](https://www.docker.com/pricing/faq/). 

Colima is an open source software that is already used by developers in this organization and has a similar set of CLI commands as Docker. 

This will satisfy issue https://github.com/cisagov/development-guide/issues/95.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

Commit:
Add Colima installation over Docker Desktop

Since our suggested containerization tool
is Colima it makes sense to update the
development guide to reflect this and remove
the Docker installation steps.

michaelsaki committed:
Fix grammar and add backticks to inline code

Commit:
Make the instructions more clear and concise

dv4harr10 commented:
Hi Michael, I noticed 2 small issues: 1) for dev_envs/mac-env-setup.md line 153 , requires a wording change from 'optionalcon figuration' to 'optional configuration'. 2) for two brew install commands I received the error: Permission denied @ apply2files-/usr/local/lib/docker/cli-plugins. (This occurred when running 'brew install docker-buildx' and 'brew install colima docker docker-compose').

michaelsaki commented:
> Hi Michael, I noticed 2 small issues: 1) for dev_envs/mac-env-setup.md line 153 , requires a wording change from 'optionalcon figuration' to 'optional configuration'. 2) for two brew install commands I received the error: Permission denied @ apply2files-/usr/local/lib/docker/cli-plugins. (This occurred when running 'brew install docker-buildx' and 'brew install colima docker docker-compose').

I think the second issue might be because you haven't removed previous docker configurations from your system.

dv4harr10 commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+
+### Optional Configurations ###
+
+Enhance your Colima setup with the following optionalcon figurations:
```
```suggestion
Enhance your Colima setup with the following optional configurations:
```

michaelsaki committed:
Fix spelling error

Co-authored-by: David Harris <123905168+dv4harr10@users.noreply.github.com>

jsf9k commented on dev_envs/mac-env-setup.md:
````diff
@@ -43,12 +43,6 @@ sudo xcodebuild -license
 xcode-select --install
 ```
 
-### Docker ###
-
-Download the latest stable [Docker for Mac](https://www.docker.com/docker-mac).
-Install it by double-clicking the downloaded `dmg` file and dragging the
-Docker application file to the `Applications` folder.
-
````
Why not leave this section?  Some folks will be OK with installing Docker for Mac.

dav3r replied:
Is our team still the target audience for this document?  If so, I don't think we need the Docker for Mac section anymore.  If we ever purchase Docker, we can always add it back.

If the target audience is wider than our team, then I agree that we can leave it, maybe adding a sentence about the licensing.

michaelsaki replied:
I was structuring it like @dav3r was suggesting. I can also see @jsf9k's point since the project is OSS and in an ideal situation outside contributions are encouraged. The question I have is are the components that they would use Docker for exclusive to CISA VPN's, AWS accounts, etc.? If that is the case then I suppose it wouldn't really matter if they had Docker or Colima if they can't really make any of the components work with it unless they just want to use it for their own sandbox environment type stuff.

michaelsaki replied:
@dav3r and @jsf9k any consensus? I am leaning towards leaving it in but just wanted to check with you all for final decision.

jsf9k replied:
I vote for leaving it in, but adding an italicized paragraph at the very beginning of the section that states _Note that [Colima](link_to_Colima_section) is our preferred solution on ~OSX~ macOS._

jsf9k replied:
Also note that the download is now called "Docker Desktop on Mac".  Note the addition of "Desktop" _and_ the modified preposition.

dav3r replied:
> I vote for leaving it in, but adding an italicized paragraph at the very beginning of the section that states _Note that [Colima](link_to_Colima_section) is our preferred solution on OSX._

I'm fine with that approach, but will note that "macOS" is the current modern nomenclature vs. "OSX".

michaelsaki replied:
@jsf9k and @dav3r I made the fix here: https://github.com/cisagov/development-guide/pull/96/commits/a7d7d8e99dc3f1bab1c450aee87e9d8bfb4c39cc

jsf9k replied:
See [this tweak suggestion](https://github.com/cisagov/development-guide/pull/96#discussion_r1497979493).

michaelsaki replied:
@jsf9k consider it done

dav3r replied:
@jsf9k - Can this conversation be resolved?

jsf9k commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
 packages. We recommend installing all the packages specified in the
 [CISA `laptop` script repository](https://github.com/cisagov/laptop/blob/master/Brewfile).
 
+### Colima: A Docker Alternative ###
```
We use sentence case for second-level headings and below:
```suggestion
### Colima: A Docker alternative ###
```

jsf9k commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+
+    The `docker-compose` usage menu should appear.
+
+### Optional Configurations ###
```
```suggestion
### Optional configurations ###
```

jsf9k commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+
+Enhance your Colima setup with the following optional configurations:
+
+#### Docker Compose as a Docker Plugin ####
```
```suggestion
#### Docker Compose as a Docker plugin ####
```

jsf9k commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+    ln -sfn $(brew --prefix)/opt/docker-buildx/bin/docker-buildx ~/.docker/cli-plugins/docker-buildx
+    ```
+
+#### Customizing VM Resources ####
````
```suggestion
#### Customizing VM resources ####
```

michaelsaki committed:
Remove capitalizations for second level headings

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+**Installation Steps:**
+
+1. Install Colima and Docker
+   command-line tools using Homebrew:
```
No need to split this line up:
```suggestion
1. Install Colima and Docker command-line tools using Homebrew:
```

dav3r commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +100,114 @@ After installing Brew, you'll want to install other useful
+    colima start
+    ```
+
+    This initializes the Docker Daemon within a lightweight VM.
````
Remove unnecessary capitalization:
```suggestion
    This initializes the Docker daemon within a lightweight VM.
```

dav3r reviewed (commented):
Not a full review, but a couple of items caught my eye.

michaelsaki committed:
Remove unnecessary capitalizations and newline

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

dav3r commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +100,113 @@ After installing Brew, you'll want to install other useful
+    docker compose
+    ```
+
+    Ensure the `docker-compose` usage menu is displayed.
````
Minor, but we might as well be consistent:
```suggestion
    Ensure the `docker compose` usage menu is displayed.
```

michaelsaki committed:
Remove hyphen for `docker compose`

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +100,113 @@ After installing Brew, you'll want to install other useful
+### Colima: A Docker alternative ###
+
+For individuals or businesses seeking a container solution without
+the licensing conditions associated with [Docker](https://www.docker.com/pricing/faq/)
```
```suggestion
the licensing conditions associated with [Docker Desktop](https://www.docker.com/pricing/faq/)
```

jsf9k reviewed (changes requested):
One very minor thang.

michaelsaki committed:
Add `Desktop` to `Docker Desktop`

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

Commit:
Revert Docker instructions and add note for Colima

jsf9k commented on dev_envs/mac-env-setup.md:
```diff
@@ -45,7 +45,10 @@ xcode-select --install
 ### Docker ###
 
-Download the latest stable [Docker for Mac](https://www.docker.com/docker-mac).
+> **Note**
+> *[Colima](https://github.com/abiosoft/colima) is our preferred solution on macOS.*
```
I think this renders a little better, and note that I changed the Colima link to point to the relevant section of this document:
```suggestion
> [!NOTE]
> [Colima](#colima-a-docker-alternative) is our preferred solution on macOS.
```

michaelsaki committed:
Improve format and update link for Colima in markdown

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r reviewed (approved):
👍 👍

Commit:
Guide users to configure Docker daemon with Colima

Adds documentation for linking Colima with the Docker daemon. This ensures
Docker client tools communicate with Colima's managed Docker daemon,
essential for Docker-based applications like Molecule without Docker Desktop.
Includes the setup code snippet and its importance for better user understanding
and usability in development workflows.

michaelsaki commented:
Something I noticed when running `Molecule` was that I left out the steps for configuring the new Docker daemon managed by Colima. Not a huge change but still thought you all would want to take a look: https://github.com/cisagov/development-guide/pull/96/commits/b8034d4558c81ed2462e233f850624c0a209502c

jsf9k commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,127 @@ After installing Brew, you'll want to install other useful
+export DOCKER_HOST=unix://$HOME/.colima/default/docker.sock
+```
+
+> **Note**
````
Prefer [the GitHub Markdown syntax for notes, warnings, cautions, etc.](https://github.com/orgs/community/discussions/16925):
```suggestion
> [!NOTE]
```

michaelsaki committed:
Change syntax for Note in MD

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,127 @@ After installing Brew, you'll want to install other useful
+    colima start --cpu 4 --memory 4 --disk 100
+    ```
+
+### Configure Colima's Docker daemon ###
+
+In your .zshrc profile add the following configuration.
+
+```console
+# Set env variable for Colima compatibility with Docker
+export DOCKER_HOST=unix://$HOME/.colima/default/docker.sock
+```
+
+> [!NOTE]
+> This configuration directs Docker client tools to communicate with the
+> Docker daemon managed by Colima, ensuring that Docker commands and
+> applications (e.g. Molecule) function as expected without Docker Desktop.
+
````
Make section heading more accurate and also add a link to the Colima FAQ:
```suggestion
### Support applications that are unaware of Docker context ###

In your .zshrc profile add the following configuration.

```console
# Set env variable for Colima compatibility with Docker
export DOCKER_HOST=unix://$HOME/.colima/default/docker.sock
```

> [!NOTE]
> This configuration directs Docker client tools to communicate with the
> Docker daemon managed by Colima, ensuring that Docker commands and
> applications (e.g. Molecule) function as expected without Docker Desktop.
> For more information, see [here](https://github.com/abiosoft/colima/blob/main/docs/FAQ.md#cannot-connect-to-the-docker-daemon-at-unixvarrundockersock-is-the-docker-daemon-running).

```

dav3r commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,127 @@ After installing Brew, you'll want to install other useful
+    colima start --cpu 4 --memory 4 --disk 100
+    ```
+
+### Configure Colima's Docker daemon ###
````
Make section heading more accurate:
```suggestion
### Support applications that are unaware of Docker context ###
```

dav3r commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,127 @@ After installing Brew, you'll want to install other useful
+> [!NOTE]
+> This configuration directs Docker client tools to communicate with the
+> Docker daemon managed by Colima, ensuring that Docker commands and
+> applications (e.g. Molecule) function as expected without Docker Desktop.
```
Add a link to the Colima FAQ:
```suggestion
> applications (e.g. Molecule) function as expected without Docker Desktop.
> For more information, see [here](https://github.com/abiosoft/colima/blob/main/docs/FAQ.md#cannot-connect-to-the-docker-daemon-at-unixvarrundockersock-is-the-docker-daemon-running).
```

michaelsaki committed:
Update heading for Docker context support

Change the section heading from "Configure Colima's Docker daemon" to "Support applications that are unaware of Docker context" to make it more accurate and reflective of the content.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

michaelsaki committed:
Add link to Colima FAQ for more info on Docker daemon

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

dav3r reviewed (approved):
Thanks for adding this! 👍 👍

jsf9k reviewed (approved):
_Approval intensifies!!!_

michaelsaki commented:
> _Approval intensifies!!!_

*Final boss approaches!* @mcdonnnj

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
 packages. We recommend installing all the packages specified in the
 [CISA `laptop` script repository](https://github.com/cisagov/laptop/blob/master/Brewfile).
 
+### Colima: A Docker alternative ###
```
I would like this whole section to be a sub-section for Docker (above). It still relies on Docker under the hood (by default) and simply provides an alternative for Docker Desktop.

michaelsaki replied:
@mcdonnnj see here https://github.com/cisagov/development-guide/pull/96/commits/8ce2503e2986f93318c97fd1ea32a649917c989f

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+For individuals or businesses seeking a container solution without
+the licensing conditions associated with [Docker Desktop](https://www.docker.com/pricing/faq/)
+, [Colima](https://github.com/abiosoft/colima) offers a compelling
+open-source alternative. Compatible with both MacOS and Linux,
```
```suggestion
open-source alternative. Compatible with both macOS and Linux,
```

michaelsaki commented:
@mcdonnnj Last call on the review for this. 🚂

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+
+Enhance your Colima setup with the following optional configurations:
+
+#### Docker Compose as a Docker plugin ####
```
So the `docker-compose` formula _is_ the Compose plugin. You can see in the [formula's code](https://github.com/Homebrew/homebrew-core/blob/277c6de80b2e5ba38b11bf0fff467c1ad8b2fdb3/Formula/d/docker-compose.rb) that it's pulling down the latest version. It just creates a `docker-compose` binary that is symlinked into the plugins directory.

michaelsaki replied:
Good catch! Verified it on my end. Everything looks good without the config stuff.

See here: https://github.com/cisagov/development-guide/pull/96/commits/8ce2503e2986f93318c97fd1ea32a649917c989f

mcdonnnj commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+1. Verify `docker-compose` is working:
+
+    ```console
+    docker-compose
````
So that we're consistent with non-macOS usage:

```suggestion
    docker compose
```

mcdonnnj commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+    docker-compose
+    ```
+
+    The `docker-compose` usage menu should appear.
````
Consistent with how we describe Docker above.
```suggestion
    The Docker Compose usage menu should appear.
```

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+
+    You should see the Docker usage menu.
+
+1. Verify `docker-compose` is working:
```
Same thing.

```suggestion
1. Verify Docker Compose is working:
```

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+
+    Ensure the `docker compose` usage menu is displayed.
+
+#### Installing Buildx ####
```
This is also already installed as a Docker plugin by the formula per the [formula's code](https://github.com/Homebrew/homebrew-core/blob/9b95aa65c9a45fea7159dcf30dd4c256493b77e4/Formula/d/docker-buildx.rb).

michaelsaki replied:
See here: https://github.com/cisagov/development-guide/pull/96/commits/8ce2503e2986f93318c97fd1ea32a649917c989f

mcdonnnj commented on dev_envs/mac-env-setup.md:
```diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+
+### Support applications that are unaware of Docker context ###
+
+In your .zshrc profile add the following configuration.
```
So we're not too specific.

```suggestion
In your shell profile (`.bashrc`, `.zshrc`, etc.) add the following configuration.
```

mcdonnnj commented on dev_envs/mac-env-setup.md:
````diff
@@ -106,6 +109,128 @@ After installing Brew, you'll want to install other useful
+
+In your .zshrc profile add the following configuration.
+
+```console
````
Since this is actual shell code.

```suggestion
```shell
```

mcdonnnj reviewed (changes requested):
Strong work 💪💪💪 I have some feedback for your consideration.

> [!NOTE]
> I tested and verified the plugin interactions with the `docker-compose` and `docker-buildx` formulae. Immediately after installation I was able to run a `docker <plugin>` and get the help output while before installation Docker complained about an unknown command.

michaelsaki committed:
Reword some of the documentation for clarity

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

michaelsaki committed:
Change "docker-compose" to "docker compose" in documentation

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Remove plugin configs and move instructions

It was discovered that the docker-compose formula is the Compose plugin.
It creates a docker-compose binary that is symlinked into the plugins
directory. This is also the case for the docker-buildx plugin. Since this
is the case it makes sense to remove the instructions for configuring the
plugins.

The Colima instructions were also moved to be a sub section under
the Docker section since Colima relies on Docker under the hood and is
simply an alternative to Docker.
